Add job controller test verifying if backoff is reseted on success

soltysh opened the pull request:
This adds a test for our job controller proving that the backoff limit is being reseted after a successful run. 

Fixes https://github.com/kubernetes/kubernetes/issues/54904

```release-note
NONE
```

soltysh committed:
Add job controller test verifying if backoff is reseted on success

soltysh commented:
@kubernetes/sig-apps-pr-reviews

soltysh commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1269,3 +1258,78 @@ func bumpResourceVersion(obj metav1.Object) {
+				{0, 2, 0, 1},
+				{0, 0, 1, 0},
+			},
+		},
```
@foxish if I add the 3rd test here, with +1 failure and +1 success this is where it won't get reseted, and I'm not 100% convinced it should. /cc @clamoriniere1A

soltysh replied:
My reasoning is that with small parallelism it should not be a problem, but what about big ones, esp. bigger than backoff limit? Or maybe we should ensure the backoff limit can't be lower than the parallelism, at least.

foxish replied:
That makes sense, maybe we should set the `backofflimit` to `min(6, parallelism)`?

foxish replied:
On second thought - maybe that's independent. With both failure and success at the same time, maybe we should leave the backoff time unchanged and retry?

soltysh replied:
I've created https://github.com/kubernetes/kubernetes/issues/56853 to discuss this and have it solved for 1.10. Here I've added a separate commit to change the completions to 3 in that e2e to get rid of that flake.

soltysh replied:
I want to leave that unit test for the backoff limit resetting, in place.

soltysh committed:
Decrease the number of completions for flaky test

soltysh commented:
@foxish ptal

foxish commented:
/lgtm
/approve

foxish commented:
Thanks @soltysh. The test will be useful for checking the behavior in future.

foxish commented:
Can you update the title of this PR @soltysh

On Dec 5, 2017 12:49 PM, "fejta-bot" <notifications@github.com> wrote:

> /retest
> This bot automatically retries jobs that failed/flaked on approved PRs
> (send feedback to @fejta).
>
> Review the full test history <https://k8s-gubernator.appspot.com/pr/56702>
> for this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/56702#issuecomment-349402596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3U59HGsc_7XyZRYdYVWzOaP-1oplA2ks5s9ZA2gaJpZM4QygSA>
> .
>

foxish commented:
/hold

@soltysh, these seem like real failures and not flakes.

soltysh commented:
> @soltysh, these seem like real failures and not flakes.

Checking...

soltysh commented:
OMG, I've dropped the completions on job definition but not on the expected line. And `framework.WaitForJobFinish` expects there a number explicitly. Updated should be good to go.

soltysh commented:
The test are green I'm gonna manually re-apply lgtm label back so that when queue is open this gets in.
